### PR TITLE
Custom SyntaxError handler.

### PR DESCRIPTION
**Pull request by Carreau**
@asmeurer would this suits, you? You have nothing to change, the `SyntaxError`s should be caughts now.

### CODE REVIEW

File: IPython/core/interactiveshell.py [hunk 1]
```diff
@@ -2682,14 +2682,18 @@ def error_before_exec(value):
...
+                    self.CustomTB(etype, value, tb)
+                    return error_before_exec(e)
                 except IndentationError as e:
                     self.showindentationerror()
                     if store_history:
                         self.execution_count += 1
                     return error_before_exec(e)
                 except (OverflowError, SyntaxError, ValueError, TypeError,
                         MemoryError) as e:
-                    self.showsyntaxerror()
```
Comment (takluyver) [inline IPython/core/interactiveshell.py]: Why's this changing? Won't it just call `showsyntaxerror` anyway?

File: IPython/core/tests/test_ultratb.py [hunk 2]
```diff
@@ -196,7 +196,7 @@ def test_non_syntaxerror(self):
             raise ValueError('QWERTY')
         except ValueError:
             with tt.AssertPrints('QWERTY'):
-                ip.showsyntaxerror()
+                ip.showtraceback()
```
Comment (takluyver) [inline IPython/core/tests/test_ultratb.py]: From the comment in this test, changing this misses the point of what it's supposed to be testing.

Comment (Carreau) [inline IPython/core/tests/test_ultratb.py]: Oops.

File: IPython/core/interactiveshell.py (hunk 1 above)
Comment (Carreau) [inline IPython/core/interactiveshell.py]: Yes it should, but it seem hard to follow the logic if some places call `syntaxerror()` and others `showtraceback()`, 
so I tried to simplify by onl calling `showtraceback()`, still I think this might be wrong. I'll revert.

**Comment by Carreau** [comment]
Fixed. Thanks for th review.

**Comment by takluyver** [comment]
Looks OK, then.
